crates/rattler_solve/src/resolvo/conda_util.rs

```diff
@@ -1,17 +1,344 @@
+use resolvo::{Dependencies, NameId, Requirement, SolvableId, SolverCache, VersionSetId};
 
 use crate::resolvo::CondaDependencyProvider;
 
+use super::SolverPackageRecord;
+
 #[derive(Copy, Clone, Debug, Eq, PartialEq)]
 pub(super) enum CompareStrategy {
     Default,
     LowestVersion,
 }
 
+/// Sorts the candidates based on the strategy.
+/// and some different rules
+pub struct SolvableSorter<'a, 'repo> {
+    solver: &'a SolverCache<CondaDependencyProvider<'repo>>,
+    strategy: CompareStrategy,
+}
+
+impl<'a, 'repo> SolvableSorter<'a, 'repo> {
+    pub fn new(
+        solver: &'a SolverCache<CondaDependencyProvider<'repo>>,
+        strategy: CompareStrategy,
+    ) -> Self {
+        Self { solver, strategy }
+    }
+
+    fn solvable_record(&self, id: SolvableId) -> &SolverPackageRecord<'repo> {
+        let pool = &self.solver.provider().pool;
+        let solvable = pool.resolve_solvable(id);
+
+        &solvable.record
+    }
+
+    /// This function can be used for the initial sorting of the candidates.
+    pub fn sort_by_name_version_build(&self, solvables: &mut [SolvableId]) {
+        solvables.sort_by(|a, b| self.simple_compare(*a, *b));
+    }
+
+    /// Sort the candidates based on:
+    /// 1. Whether the package has tracked features
+    /// 2. The version of the package
+    /// 3. The build number of the package
+    fn simple_compare(&self, a: SolvableId, b: SolvableId) -> Ordering {
+        let a_record = &self.solvable_record(a);
+        let b_record = &self.solvable_record(b);
+
+        // First compare by "tracked_features". If one of the packages has a tracked
+        // feature it is sorted below the one that doesn't have the tracked feature.
+        let a_has_tracked_features = !a_record.track_features().is_empty();
+        let b_has_tracked_features = !b_record.track_features().is_empty();
+        match a_has_tracked_features.cmp(&b_has_tracked_features) {
+            Ordering::Less => return Ordering::Less,
+            Ordering::Greater => return Ordering::Greater,
+            Ordering::Equal => {}
+        };
+
+        // Otherwise, select the variant with the highest version
+        match (self.strategy, a_record.version().cmp(b_record.version())) {
+            (CompareStrategy::Default, Ordering::Greater)
+            | (CompareStrategy::LowestVersion, Ordering::Less) => return Ordering::Less,
+            (CompareStrategy::Default, Ordering::Less)
+            | (CompareStrategy::LowestVersion, Ordering::Greater) => return Ordering::Greater,
+            (_, Ordering::Equal) => {}
+        };
+
+        // Otherwise, select the variant with the highest build number
+        match a_record.build_number().cmp(&b_record.build_number()) {
+            Ordering::Less => return Ordering::Greater,
+            Ordering::Greater => return Ordering::Less,
+            Ordering::Equal => return Ordering::Equal,
+        };
+    }
+
+    pub fn sort_by_dependencies(
```

Comment:
The API of this function is maybe slight weird because it assumes that the input is already sorted by `sort_by_name_version_build`. Would it make sense to have one `pub` function in this `impl` that calls both function in succession and make the other functions not-pub? Or have `sort_by_name_version_build` return a wrapper type that guarentees the sorted properties that is consumed by this function?

Reply:
Yes I'll refactor.